r/R/dplyr-slice.R

```diff
@@ -0,0 +1,144 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+
+# The following S3 methods are registered on load if dplyr is present
+
+slice_head.arrow_dplyr_query <- function(.data, ..., n, prop) {
+  if (length(group_vars(.data)) > 0) {
+    arrow_not_supported("Slicing Arrow data with groups")
+  }
+
+  if (missing(n)) {
+    n <- prop_to_n(.data, prop)
+  }
+
+  head(.data, n)
+}
+slice_head.Dataset <- slice_head.ArrowTabular <- slice_head.RecordBatchReader <- slice_head.arrow_dplyr_query
+
+slice_tail.arrow_dplyr_query <- function(.data, ..., n, prop) {
+  if (length(group_vars(.data)) > 0) {
+    arrow_not_supported("Slicing Arrow data with groups")
+  }
+
+  if (missing(n)) {
+    n <- prop_to_n(.data, prop)
+  }
+
+  tail(.data, n)
+}
+slice_tail.Dataset <- slice_tail.ArrowTabular <- slice_tail.RecordBatchReader <- slice_tail.arrow_dplyr_query
+
+slice_min.arrow_dplyr_query <- function(.data, ..., n, prop, with_ties = TRUE) {
+  if (length(group_vars(.data)) > 0) {
+    arrow_not_supported("Slicing Arrow data with groups")
+  }
+  if (with_ties) {
+    arrow_not_supported("with_ties = TRUE")
+  }
+
+  if (missing(n)) {
+    n <- prop_to_n(.data, prop)
+  }
+
+  head(dplyr::arrange(.data, ...), n)
+}
+slice_min.Dataset <- slice_min.ArrowTabular <- slice_min.RecordBatchReader <- slice_min.arrow_dplyr_query
+
+slice_max.arrow_dplyr_query <- function(.data, ..., n, prop, with_ties = TRUE) {
+  if (length(group_vars(.data)) > 0) {
+    arrow_not_supported("Slicing Arrow data with groups")
+  }
+  if (with_ties) {
+    arrow_not_supported("with_ties = TRUE")
+  }
+
+  if (missing(n)) {
+    n <- prop_to_n(.data, prop)
+  }
+
+  sorted <- dplyr::arrange(.data, ...)
+  # Invert the sort order of the things in ... so they're descending
+  # TODO: handle possibility that .data was already sorted and we don't want
+  # to invert those sorts? Does that matter? Or no because there's no promise
+  # of order of which TopK elements you get if there are ties?
+  sorted$arrange_desc <- !sorted$arrange_desc
+  head(sorted, n)
+}
+slice_max.Dataset <- slice_max.ArrowTabular <- slice_max.RecordBatchReader <- slice_max.arrow_dplyr_query
+
+slice_sample.arrow_dplyr_query <- function(.data,
+                                           ...,
+                                           n,
+                                           prop,
+                                           weight_by = NULL,
+                                           replace = FALSE) {
+  if (length(group_vars(.data)) > 0) {
+    arrow_not_supported("Slicing Arrow data with groups")
```

Comment:
```suggestion
    arrow_not_supported("Slicing grouped data")
```
Slightly more concise - feel free to disregard